### ISSUE TITLE
About gradient reduction in DistributionStrategy

### ISSUE THREAD

**Issue by wangsiyu** [tensorflow/tensorflow#25806]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source 
- TensorFlow version (use command below): master
- Python version: 2.3.7
- Bazel version (if compiling from source): 0.2.0
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: cuda 9.0, cudnn 7
- GPU model and memory:

**Describe the current behavior**
1. Currently the gradient reduction in `distributed_apply` is `SUM` not `MEAN`. This makes learning rate adjustment more confusing. In general, learning rate should be adjusted according to total batch size. If it is `SUM` case, the learning rate adjustment will be different between single mode and distribution mode given the same total batch size.  

2. The gradient reduction is implemented in `_distributed_apply` function. That means the gradient clipping will be done before gradient reduction. This is not identical with single mode because the gradient clipping should be done after reduction along the `batch size` dimenstion. Another, in `clip_by_global_norm` case, it blocks the overlapping between computation and communication. The performance is even worse then using `CollectiveAllReduceStrategy`. I have tried to put the gradient reduction in `def gradients` instead of `_distributed_apply` locally. If you think this is necessary, I will refine my code and submit a pull request here.
 
**Describe the expected behavior**
1. Change the gradient reduction from `SUM` to `MEAN`
2. Move gradient reduction from `_distributed_apply` to `def gradients`

**Other info / logs**
I will contribute for it if you think the above are necessary. Thanks. @yuefengz

**Comment by yuefengz** [comment]
For 1), the gradient reduction is `SUM` because there is no nccl mean. Using `SUM` avoids the extra division for all gradients. We actually [scaled down the loss](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/losses/losses_impl.py#L195) on each replica so that the gradient reduction is effectively `MEAN`.

For 2), let me think about it. @guptapriya can also comment on it.

**Comment by wangsiyu** [comment]
For 1), I have some questions to ask.
1. For scaling down the loss, There must be another two conditions for taking it into effect.  a) To build loss function by using `tf.losses.softmax_cross_entropy` instead of `tf.nn.softmax_cross_entropy`. b) [`reduction](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/losses/losses_impl.py#L722)` params should be set to `MEAN` from outside, right?
2. `Estimator` will [do `SUM` for loss](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/estimator.py#L1270). So it just have effects on displaying the loss, right?

**Comment by fanshiqing** [comment]
@yuefengz For 1), it seems that [loss is only scaled down](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/estimator.py#L1300-L1328) for display's convenience, while at this time the replica [training graph has been built](https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/estimator.py#L1294-L1299), which means each replica will use their own loss at backward pass.  Hence in my opinion the gradient is still `SUM` but not `MEAN`.

> For 1), the gradient reduction is `SUM` because there is no nccl mean. Using `SUM` avoids the extra division for all gradients. We actually [scaled down the loss](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/losses/losses_impl.py#L195) on each replica so that the gradient reduction is effectively `MEAN`.
> 
> For 2), let me think about it. @guptapriya can also comment on it.

**Comment by guptapriya** [comment]
Loss scaling happens in different places depending on which high level API and optimizer you're using. If you're using estimator with TF 1 optimizer, loss scaling happens in [optimizer.compute_gradients](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py#L468). In 2.0 when using Keras + TF 2 optimizer, the scaling happens inside Keras instead of Optimizer. 

@fanshiqing the code you linked to about loss is not scaling it, it is just aggregating the loss from different replicas. By this point, the loss from each replica has already been scaled (by the number of replicas). 

@wangsiyu - yes the loss reduction type must be MEAN for any loss scaling to happen. If loss reduction type is SUM, we will just be doing SUM of gradients across replicas as well (which matches what you would do on one replica as well)

To recap, we (1) scale the loss on each replica by dividing by number of replicas, (2) compute gradients on each replica (3) sum the gradients from each replica (4) add the loss from each replica for reporting purpose. So effectively, we (1) "Average gradients" (because scaling the loss has the same effect as scaling the gradients) and (2) "Average the loss across replicas for reporting". 

RE: gradient clipping, I don't understand why it should only happen after the reduction across replicas. You mention "gradient clipping should be done after reduction along the batch size dimension". But gradients do not have a batch dimension, so there is no reduction along batch dimension. The reduction is only across replicas.

**Comment by wangsiyu** [comment]
Thanks for your reply. 
For gradient reduction, I did not notice the loss scale in `compute_gradients` function. I have a litter suggestion. You can move the loss scale from `compute_gradients` to `tf.gradients`. Because some developers still use `tf.gradients` instead of that one.
For gradient clipping, my statement is somewhat ambiguous. The key point is reduction after clipping is not equivalent with clipping after reduction. For example, when using global clipping, the global norm is not the same in these two cases. Another, reduction after global clipping is not conducive to the overlap of computation and communication.

**Comment by guptapriya** [comment]
For 2.0, as I mentioned, we've removed the scaling from compute_gradients. Now the scaling is left up to the user if using custom training loops. If you're using estimator in TF 2.0 also, you need to do the scaling in your model_fn yourself. 
As for gradient clipping, I think there is no one right answer. If you prefer to do the clipping after reduction, you can override the optimizer's `apply_gradients` call perhaps?

**Comment by nikitamaia** [comment]
Closing this issue since there is no bug here.
